Sources/AsyncHTTPClient/FileDownloadDelegate.swift

```diff
@@ -0,0 +1,88 @@
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+/// Handles a streaming download to a given file path, allowing headers and progress to be reported.
+public final class FileDownloadDelegate: HTTPClientResponseDelegate {
+    /// The response type for this delegate: the total count of bytes as reported by the response
+    /// "Content-Length" header (if available) and the count of bytes downloaded.
+    public typealias Response = (totalBytes: Int?, receivedBytes: Int)
+
+    private var totalBytes: Int?
+    private var receivedBytes = 0
+
+    private let handle: NIOFileHandle
+    private let io: NonBlockingFileIO
+    private let reportHeaders: ((HTTPHeaders) -> Void)?
+    private let reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)?
+
+    private var writeFuture: EventLoopFuture<Void>?
+
+    /// Initializes a new file download delegate.
+    /// - parameters:
+    ///     - path: Path to a file you'd like to write the download to.
+    ///     - pool: A thread pool to use for asynchronous file I/O.
+    ///     - reportHeaders: A closure called when the response headers are available.
+    ///     - reportProgress: A closure called when a body chunk has been downloaded, with
+    ///       the total byte count and download byte count passed to it as arguments.
+    public init(
+        path: String,
+        pool: NIOThreadPool = NIOThreadPool(numberOfThreads: 1),
+        reportHeaders: ((HTTPHeaders) -> Void)? = nil,
+        reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)? = nil
+    ) throws {
+        self.handle = try NIOFileHandle(path: path, mode: .write, flags: .allowFileCreation())
+        pool.start()
+        self.io = NonBlockingFileIO(threadPool: pool)
+
+        self.reportHeaders = reportHeaders
+        self.reportProgress = reportProgress
+    }
+
+    public func didReceiveHead(
+        task: HTTPClient.Task<Response>,
+        _ head: HTTPResponseHead
+    ) -> EventLoopFuture<Void> {
+        self.reportHeaders?(head.headers)
+
+        if let totalBytesString = head.headers.first(name: "Content-Length"),
+            let totalBytes = Int(totalBytesString) {
+            self.totalBytes = totalBytes
+        }
+
+        return task.eventLoop.makeSucceededFuture(())
+    }
+
+    public func didReceiveBodyPart(
+        task: HTTPClient.Task<Response>,
+        _ buffer: ByteBuffer
+    ) -> EventLoopFuture<Void> {
+        self.receivedBytes += buffer.readableBytes
+        self.reportProgress?(self.totalBytes, self.receivedBytes)
+
+        let writeFuture = self.io.write(fileHandle: self.handle, buffer: buffer, eventLoop: task.eventLoop)
+        self.writeFuture = writeFuture
+        return writeFuture
+    }
+
+    public func didFinishRequest(task: HTTPClient.Task<Response>) throws -> Response {
+        self.writeFuture?.whenComplete { [weak self] _ in
+            try! self?.handle.close()
```

Comment:
I don't think we should crash if the `close` fails

Reply:
I've made it `try!` based on this suggestion from @weissi in the Swift Server Slack (`#async-http-client` channel):

>and you want to `try!` the `close.` The only reason for the `close` to fail is that you have already closed it. If you have already closed it, then that's a serious programmer error that need fixing.